src/builtins.ts

```diff
@@ -137,6 +137,15 @@ export function compileCall(
         classType !== null && classType.prototype.extends(compiler.program.arrayPrototype)
       ) ? module.createI32(1) : module.createI32(0);
     }
+    case "isArrayLike": { // isArrayLike<T!>() / isArrayLike<T?>(value: T) -> bool
+      let type = evaluateConstantType(compiler, typeArguments, operands, reportNode);
+      compiler.currentType = Type.bool;
+      if (!type) return module.createUnreachable();
+      let classType = type.classReference;
+      return (
+        classType !== null && classType.lookupFieldOrProperty("length")
```

Comment:
Another thing we might need to check here is to check that it is of type `i32`. While straightforward for fields, for properties, that's essentially the return type of the getter.

Reply:
I think it will be any integer number (i8/u16/i32/u64). Could you take tip how this do properly?

Reply:
As a check, you could use `type.isAssignableTo(Type.i32)`.

Reply:
@dcodeIO still can't figure out how this provide

Reply:
Yeah, seems to be somewhat suboptimal in the current IR. `Property` is more like a meta element that groups two prototypes, but doesn't provide anything to access the readily resolved functions so one could easily obtain the return type. Ideally there'd also be a `PropertyInstance` for this. Would be even better if every element had a `.type`, regardless of `.kind`.

Reply:
Could we merge this without strict type checking and enhance later after refactoring? Anyway this very specific builtiin and would using mostly internally. But this unblock implementation for `Array#from` without iterators yet

Reply:
Btw. isn't this missing checks for indexed get currently?

Reply:
in TS following code is legal:
```ts
function foo(arr: ArrayLike<number>): number {
  return arr.length;
}
foo({ length: 10 });
```
Or
```ts
Array<number>.from({ length: 10 });
```
So we could declare this as:
```ts
class Array<T> {
  from<T, U>(arrayLike: ArrayLike<T>, mapfn: (val: T, key: i32) => U): Array<U>;
  // from<T>(arrayLike: ArrayLike<T>): Array<T>;
}
```
In TS any object implicitly has dictionary interface. So `{ length: 10 }` could be valid cast to `ArrayLike`. How this resolve for AS classes/objects I don't know but my goal is support `Array<i32>.from({ length: 100 })` like this in AS as well.

Reply:
Could I solve this puzzle after resolve refactoring? =)

Reply:
Still not quite confident that `Array.from({ length: 10 })` is going to work out as intended. One thing I can imagine `isArrayLike` being especially useful for is if a (standard library) function needs to work with both arrays and strings, and since we have indexed overloads on strings as well, this would "just work". This could also be easily extended by a user to work with any custom class under the condition that an indexed get overload is present. TL;DR Without a check for indexed get as well this seems less useful to me.

Reply:
I suggest add build-in which check is index accessor (operator) present or not. So you can use restricted checks like ``isArrayLike<T>() && hasIndexableOperator<T>()`. WDYT?